### DISCUSSION TITLE
Sprint 2 Retrospective

### DISCUSSION THREAD

**Discussion by ethanfrech** [cs481-ekh/f22-softskills#144] [Sprint Retrospective]
# Sprint 2 Retrospective #

Overall Sprint 2 proved to be a difficult yet productive endeavor. As a team we experienced individual challenges that hampered our collective progress to degree. Fortunately after communicating our respective struggles we were able to formulate and commit to a decisive course of action that accommodated everyone. 

We decided as a team that instead of using Sveltekit for our application we would switch to express, JavaScript, and Typescript. While Svelte in the right hands proves to be an efficient and optimized JavaScript experience there are far few examples of how to implement common functionalities as opposed to pure JavaScript. This was especially apparent when it came to authenticating users through Google. Svelte has a generic library that can deal with json web tokens, but a solution like Passport comes with strategies for different authentication strategies out of the box. In our case since we were authenticating with Google we implemented the oauth2 strategy on our Express framework Node.js server.

This pivot only occurred more than three quarters of the way into the Sprint which effectively prevented the API core from be tested. Once we got a mock up running with Express and Passport we were able to test the API core, and were able to successfully make calls to the Google Drive API. We were able to enumerate the entire file structure of a given drive, as well as pull all relevant information from the metadata. In our case we were most interested in the permissions attribute given we are building a permission manager.

The switch from Svelte to JavaScript heavily impacted the Front-end, so UI development will be a top priority in the next Sprint. We met with our sponsor and he agrees that is what he would like to see at the end of next Sprint. This decision was informed by the fact that the Back-end is almost at MVP, so we should be able to polish it off and dedicate more time and energy to the user interface experience. 

Other than some mild interference with the Front-end, the transition away from Svelte was not conducive to a major set back. Much of the Back-end was already being developed in TypeScript. As it turned out TypeScript was also ideal for the database. The web server and authentication components spun up quickly once we incorporated Express and Passport. No integral components of the project had been written in Svelte at the time of the change. We did have to refactor the the repository to the proper file structure for the new components. The Dockerfile aslso had to be updated of course. 

We have a clear set of directives and references for the visual representations of the folders and files of the logged in user.

Next sprint there are some things we intend on improving on: 
- We want all of our tasks that involve pushing code to the repository to have a pull request for traceability.  
- We want to get our individual work done earlier in the sprint so there is ample time left to integrate the respective pieces.
- We want to communicate progress and ask for help more often to ensure cohesive, efficient development.

**************************************************************************************************************************************

## What went well: ##
- Backend is almost at MVP
    - API core and database
- Communication
    - Meetings were consistent and productive
- Better GitHub workflow understanding and usage
- Typescript introduction

## What didn't go well: ##
- Google OAuth took too long
    - API core was unable to be fully tested because of this
- Isaac stumbled working with Svelte in first half
    - Got less done than he wanted to
- Spent a lot of time refactoring to accomodate tooling change

## Events: ##
- Tooling changed from Sveltekit + Lucia to Express + Passport
- Introduced Typescript into project

## Hopeful improvements: ##
- Continue honing Git/GitHub workflow
    - Better but not perfect
- Get our work done sooner rather than later
    - Set hard deadline for individual work
    - Have several days before end of sprint to integrate different part
- Communication about work progress/asking for help
- Improve efficiency of working time

**Comment by bsu-jacksonmorton** [comment]
#79 Is waiting on #95 to be complete and will be moved to Sprint 3. Philip and I are going to coordinate during Sprint 3 to create functions for making queries to our DB for specific queries such as getting permissions by email.
